### DISCUSSION TITLE
How to control the line-breaking behavior of raw text?

### DISCUSSION THREAD

**Discussion by Evan-Zhao** [typst/typst#3988] [Q&A] [answered]
I'm writing a tutorial for a Python package and have a lot of `pkg.pkg.pkg.class.function` kind of text in the document. I'm also using `set par(justify: true, linebreaks: "optimized")` to get a Latex-like look where the ending of lines are lined up.

It turns out these two really don't play well together, because Typst doesn't want to break line midway of raw text except at spaces, as it seems. See the following example for the messed-up line spacing. 
![image](https://github.com/typst/typst/assets/19525332/9955b5fe-1968-454d-a5c1-a1421091c6f7)

However, in my case I would like Typst to know that it's allowed break lines after a dot (`.`). Is there anyway to do so, or other ways to just improve the layout in such cases?

**Comment by laurmaedje** [comment, marked as answer]
You can add explicit linebreak opportunities by adding `sym.zws` (the Unicode Zero Width Space) after the dots. You could do this automatically with a show rule like this:
```typ
#show raw.where(block: false): it => {
  show ".": "." + sym.zws
  it
}
```
